zebra-chain/src/transaction/serialize.rs

```diff
@@ -245,6 +249,123 @@ impl ZcashDeserialize for Option<sapling::ShieldedData<SharedAnchor>> {
+            }
+        }
+        Ok(())
+    }
+}
+
+impl ZcashSerialize for orchard::ShieldedData {
+    fn zcash_serialize<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        // Split the AuthorizedAction
+        let (actions, sigs): (Vec<orchard::Action>, Vec<Signature<SpendAuth>>) = self
+            .actions
+            .iter()
+            .cloned()
+            .map(orchard::AuthorizedAction::into_parts)
+            .unzip();
+
+        // nActionsOrchard and vActionsOrchard
+        actions.zcash_serialize(&mut writer)?;
+
+        // flagsOrchard
+        writer.write_u8(self.flags.bits())?;
+
+        // valueBalanceOrchard
+        self.value_balance.zcash_serialize(&mut writer)?;
+
+        // anchorOrchard
+        writer.write_all(&<[u8; 32]>::from(self.shared_anchor)[..])?;
+
+        // sizeProofsOrchard and proofsOrchard
+        self.proof.zcash_serialize(&mut writer)?;
+
+        // vSpendAuthSigsOrchard
+        zcash_serialize_external_count(&sigs, &mut writer)?;
+
+        // bindingSigOrchard
+        writer.write_all(&<[u8; 64]>::from(self.binding_sig)[..])?;
+
+        Ok(())
+    }
+}
+
+// we can't split ShieldedData out of Option<ShieldedData> deserialization,
+// because the counts are read along with the arrays.
+impl ZcashDeserialize for Option<orchard::ShieldedData> {
+    fn zcash_deserialize<R: io::Read>(mut reader: R) -> Result<Self, SerializationError> {
+        // nActionsOrchard and vActionsOrchard
+        let actions: Vec<orchard::Action> = (&mut reader).zcash_deserialize_into()?;
+
+        if actions.is_empty() {
+            // read another compactsize and get out
+            let _ = reader.read_compactsize()?;
+            return Ok(None);
+        }
+
+        // flagsOrchard
+        let flags: orchard::Flags = orchard::Flags::from_bits(reader.read_u8()?).unwrap();
+
+        // valueBalanceOrchard
+        let value_balance: amount::Amount = (&mut reader).zcash_deserialize_into()?;
+
+        // anchorOrchard
+        let shared_anchor: orchard::tree::Root = reader.read_32_bytes()?.into();
+
+        // sizeProofsOrchard and proofsOrchard
+        let proof: Halo2Proof = (&mut reader).zcash_deserialize_into()?;
+
+        // vSpendAuthSigsOrchard
+        let sigs: Vec<Signature<SpendAuth>> =
+            zcash_deserialize_external_count(actions.len(), &mut reader)?;
+
+        // bindingSigOrchard
+        let binding_sig = reader.read_64_bytes()?.into();
+
+        // Create the AuthorizedAction
+        let mut authorized_action = Vec::<orchard::AuthorizedAction>::with_capacity(actions.len());
+        for (count, action) in actions.iter().cloned().enumerate() {
+            authorized_action.push(orchard::AuthorizedAction {
+                action,
+                spend_auth_sig: sigs[count],
+            })
+        }
+        let actions: AtLeastOne<orchard::AuthorizedAction> = authorized_action.try_into()?;
```

Comment:
This can probably be improved.

Reply:
Use zip and `AuthorizedAction { field, field }`, here's an example:
https://github.com/ZcashFoundation/zebra/blob/2ea1d8624d1d72abdf582953f4f1133efe9eea4f/zebra-chain/src/transaction/serialize.rs#L217-L222

Reply:
Thanks. https://github.com/ZcashFoundation/zebra/pull/2116/commits/96d84d8461e55401427c46cb6d38c96e5d25cd00